### PR TITLE
internal/charmstore: implement multi-series find-id logic

### PR DESCRIPTION
**Pull request by rogpeppe**
This implements the behaviour described in https://docs.google.com/document/d/1hiQ1aka40Jx9QdnXHXgtGzAZopIuAkDIO-pV-cZhASc .

The EntitiesQuery logic could be made more efficient but I've left it as is because I think it more clear this way, and the actual performance hit from the redundant condition checks is likely to be negligible.

Note that this also introduces a behaviour change into the API. A charm id with a revision but no specified series will not match old charms. If anyone was using that functionality previously, they were almost certainly doing it wrong.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1191/
Test PASSed.

File: internal/charmstore/store.go [hunk 1]
```diff
@@ -561,14 +561,17 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
...
 //
 // It is the responsibility of the caller to set e.SupportedSeries
 // if the entity URL does not contain a series. If the entity
 // URL *does* contain a series, e.SupportedSeries will
 // be overwritten.
-func denormalizeEntity(e *mongodoc.Entity) {
+//
+// This is exported for the purposes of tests that
+// need to create directly into the database.
+func DenormalizeEntity(e *mongodoc.Entity) {
```
Comment (urosj) [inline internal/charmstore/store.go:574]: Hm, do we really want to expose this?

File: internal/charmstore/store.go [hunk 2]
```diff
@@ -719,40 +723,92 @@ var seriesScore = map[string]int{
...
+		return entities.Find(bson.D{{
+			"name", url.Name,
+		}, {
+			"user", url.User,
+		}, {
+			"supportedseries", url.Series,
+		}})
+	case url.User != "" && url.Revision != -1:
+		return entities.FindId(url)
+	case url.User != "":
```
Comment (urosj) [inline internal/charmstore/store.go]: I guess it's safe to not check that name is not empty as that's invalid entity.

File: internal/charmstore/store_test.go [hunk 3]
```diff
@@ -1765,6 +1766,187 @@ func (s *StoreSuite) TestFindBestEntity(c *gc.C) {
...
+		SupportedSeries: []string{"precise"},
+	}, {
+		URL:             charm.MustParseReference("~charmers/wordpress-13"),
+		SupportedSeries: []string{"trusty"},
+	}, {
+		URL:             charm.MustParseReference("~bob/wordpress-14"),
+		SupportedSeries: []string{"precise"},
+	}},
+	url:       "~charmers/precise/wordpress",
+	expectURL: "~charmers/wordpress-12",
```
Comment (urosj) [inline internal/charmstore/store_test.go]: Shouldn't this be ~charmers/wordpress-13 ?

File: internal/charmstore/store.go (hunk 1 above)
Comment (rogpeppe) [inline internal/charmstore/store.go:574]: As long as there are tests outside of internal/charmstore that rely on writing entities to the database directly, I think it's reasonable to do so. We're not exposing it to the outside world after all.

File: internal/charmstore/store.go (hunk 2 above)
Comment (rogpeppe) [inline internal/charmstore/store.go]: Yup.

**Comment by urosj** [comment]
:+1:

File: internal/charmstore/store.go (hunk 1 above)
Comment (mhilton) [inline internal/charmstore/store.go:574]: I think this should probably only be exported in an export_test.go or similar.

File: internal/charmstore/store.go [hunk 4]
```diff
@@ -673,6 +676,7 @@ func (s *Store) FindEntities(url *charm.Reference, fields ...string) ([]*mongodo
 // populated in the returned entities. If the given URL has no user then
 // only promulgated entities will be queried.
 func (s *Store) FindBestEntity(url *charm.Reference, fields ...string) (*mongodoc.Entity, error) {
+	logger.Infof("FindBestEntity %s", url)
```
Comment (mhilton) [inline internal/charmstore/store.go]: hmm, I'm not sure that this adds much value.

File: internal/charmstore/store.go [hunk 5]
```diff
@@ -719,40 +723,92 @@ var seriesScore = map[string]int{
 	"utopic":  4,
 	"vivid":   5,
 	"wily":    6,
+	// When we find a multi-series charm (no series) we
+	// will always choose it in preference to a series-specific
+	// charm
+	"": 5000,
```
Comment (mhilton) [inline internal/charmstore/store.go:728]: MaxInt or something?

File: internal/charmstore/store.go [hunk 6]
```diff
@@ -719,40 +723,92 @@ var seriesScore = map[string]int{
...
 // the produced query will only match promulgated entities.
 func (s *Store) EntitiesQuery(url *charm.Reference) *mgo.Query {
-	if url.User != "" && url.Series != "" && url.Revision != -1 {
-		// Find a specific owned entity, for instance ~who/utopic/django-42.
-		return s.DB.Entities().FindId(url)
-	}
-	if url.Series != "" && url.Revision != -1 {
-		// Find a specific promulgated entity, for instance utopic/django-42.
+	entities := s.DB.Entities()
+	// TODO this could be phrased more efficiently.
```
Comment (mhilton) [inline internal/charmstore/store.go]: This seems unnecessarily hard to follow. Which I expect is on the basis of some kind of theoretical optimizations. How about: 

``` go
query := bson.D{{
    "name", url.Name,
}}
if url.User == "" {
    if url.Revision > -1 {
        query = append(query, bson.DocElem{"promulgated-revision", url.Revision})
    } else {
        query = append(query, bson.DocElem{"promulgated-revision", bson.D{{"$gt", -1}}})
    }
} else {
    query = append(query, bson.DocElem{"user", url.User})
    if url.Revision > -1 {
        query = append(query, bson.DocElem{"revision", url.Revision})
    }
}
if url.Series == "" {
    query = append(query, bson.DocElem{"series", ""})
} else {
    query = append(query, bson.DocElem{"supportedseries", url.Series})
}
return entities.Find(query)
```

File: internal/v4/api_test.go [hunk 7]
```diff
@@ -1345,8 +1345,12 @@ var serveExpandIdTests = []struct {
 		{Id: "cs:precise/haproxy-1"},
 	},
 }, {
-	about: "single result",
+	about: "revision with series doesn't match anything",
```
Comment (mhilton) [inline internal/v4/api_test.go]: without series? or with empty series.

File: internal/v4/archive_test.go [hunk 8]
```diff
@@ -1448,52 +1437,37 @@ func (s *ArchiveSuite) TestGetNewPromulgatedRevision(c *gc.C) {
...
-			Series:              "trusty",
-			Revision:            i,
-			PromulgatedURL:      pid,
-			PromulgatedRevision: i,
-		})
+		entity := &mongodoc.Entity{
+			URL:            id,
+			PromulgatedURL: pid,
+		}
+		charmstore.DenormalizeEntity(entity)
```
Comment (mhilton) [inline internal/v4/archive_test.go:1469]: Hmm, I think this is an unreasonable usecase to export Denormalise entity for. This test should use other store methods to add the charms it needs.

**Comment by mhilton** [comment]
I don't think I'm quite +1 on this yet, The EntitiesQuery function is so much more convoluted than seems to be needed I'd like to have a chat about it before it lands.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1193/
Test PASSed.

File: internal/v4/archive_test.go (hunk 8 above)
Comment (rogpeppe) [inline internal/v4/archive_test.go:1469]: If it's OK, can we leave off fixing this test for a separate PR?

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1194/
Test PASSed.

**Comment by mhilton** [comment]
LGTM with DenormalizeEntity no longer exported, There is certainly a better way to make that test work.

File: internal/charmstore/store.go (hunk 5 above)
Comment (rogpeppe) [inline internal/charmstore/store.go:728]: Given that we've got a known set of integers enumerated here, maxint seems like overkill to me.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1195/
Test PASSed.
